compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  std::string gen;
+
+  // On pack, target buffer is input parameter. On unpack, the temporary.
+  const char* target = out ? "obj" : "result";
+
+  // Move the raw bytes of the type to/from the wire.
+  gen += "skt_err=";
+  gen += this->genSocketCall("skt", target, out);
+
+  // Generate a null frame in the opposite direction for the ACK.
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  return gen;
+}
+
+//
+//
+//
+std::string MLIContext::genMarshalBodyStringC(Type* t, bool out) {
+  const char* target = out ? "obj" : "result";
+  std::string gen;
+
+  // Read the string length before pushing.
+  if (out) { gen += "bytes = strlen(obj);\n"; }
+
+  // Push/pull string length along with null frame for ACK.
+  gen += this->genSocketCall("skt", "bytes", out);
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  if (out) {
+    // Next push this string across the wire.
+    gen += this->genSocketCall("skt", "obj", "bytes", out);
+  } else {
+    // Allocate buffer for string (include NULL terminator).
+    gen += "result = malloc(bytes + 1);\n";
+    gen += this->genTodo("Assert should shutdown server or return NULL.");
+    gen += "assert(result);\n";
+  }
+
+  // Move the string to/from the buffer.
+  gen += this->genSocketCall("skt", target, "bytes", out);
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  // Null terminate if receiving.
+  if (not out) { gen += "result[bytes] = 0;\n"; }
+
+  return gen;
+}
+
+std::string MLIContext::genMarshalRoutine(Type* t, bool out) {
+  int64_t id = this->assignUniqueTypeID(t);
+  std::string gen;
+
+  // Push returns nothing, while pull returns the type being read in.
+  if (out) {
+    gen += "void ";
+  } else {
+    gen += this->genTypeName(t);
+    gen += " ";
+  }
+
+  // Select appropriate prefix for function name based on direction.
+  gen += out ? marshal_push_prefix : marshal_pull_prefix;
+  gen += std::to_string(id);
+  gen += "(void* skt";
+
+  // Push routines expect the type as a parameter (named "obj").
+  if (out) {
+    gen += ",";
+    gen += this->genTypeName(t);
+    gen += " obj";
+  }
+
+  gen += ")";
+  gen += scope_begin;
+
+  // Always declare a variable to catch socket errors.
+  gen += this->genNewDecl("int", "skt_err");
+
+  // If unpacking, declare a temporary for the return value.
+  if (not out) {
+    gen += this->genNewDecl(t, "result");
+  }
+
+  // If allocating, declare a temporary for amount.
+  if (this->typeRequiresAllocation(t)) {
+    gen += this->genNewDecl("uint64_t", "bytes");
+  }
+
+  // Insert a debug message if appropriate.
+  if (this->injectDebugPrintlines) {
+    std::string msg;
+
+    msg += out ? "Pushing type: " : "Pulling type: ";
+    msg += this->genTypeName(t);
+
+    gen += this->genDebugPrintCall(msg.c_str(), "[RPC]");
+  }
+
+  //
+  // Handle translation of different type classes here. Note that right now
+  // the only things we can translate are primitive scalars and cstrings.
```

Comment:
I thought we couldn't handle c_strings?

Reply:
I merged a partially completed WIP branch where I was working on c_strings into this one (thus the comment).

Reply:
Code has been removed.